src/Linters/NoFinalMethodInFinalClassLinter.hack

```diff
@@ -0,0 +1,108 @@
+
+final class NoFinalMethodInFinalClassLinter extends AutoFixingASTLinter {
+  const type TNode = MethodishDeclaration;
+  const type TContext = ClassishDeclaration;
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    ClassishDeclaration $class,
+    MethodishDeclaration $method,
+  ): ?ASTLintError {
+    $class_modifiers = $class->getModifiers() ?? new NodeList();
+    if (!self::hasFinalModifier($class_modifiers)) {
+      return null;
+    }
+    $function_modifiers = $method->getFunctionDeclHeader()->getModifiers() ??
+      new NodeList();
+    if (!self::hasFinalModifier($function_modifiers)) {
+      return null;
+    }
+
+    return new ASTLintError(
+      $this,
+      Str\format(
+        'Method %s is final in the class %s, which is also final. This is redundant.',
+        Str\trim($class->getName()->getCode()),
+        Str\trim($method->getFunctionDeclHeader()->getName()->getCode()),
+      ),
+      $method,
+      () ==> $this->getFixedNode($method),
+    );
+
+    echo $class->getName()->getCode().
+      ' is final and has the final method '.
+      $method->getFunctionDeclHeader()->getName()->getCode().
+      \PHP_EOL;
+  }
+
+  <<__Override>>
+  public function getTitleForFix(LintError $_): string {
+    return 'Remove final from methods in final classes';
+  }
+
+  public function getFixedNode(
+    MethodishDeclaration $method,
+  ): MethodishDeclaration {
+    $function_decl_header = $method->getFunctionDeclHeader();
+    $modifiers = $function_decl_header->getModifiers() as nonnull;
+
+    $final = $modifiers->filterChildren($modifier ==> $modifier is FinalToken)
+      ->getFirstTokenx();
+    $non_final = $modifiers->filterChildren(
+      $modifier ==> !$modifier is FinalToken,
+    );
+
+    $decl_without_final = $function_decl_header->withModifiers($non_final);
+
+    $decl_without_final = self::preserveTrivia(
+      $final,
+      $decl_without_final->getFirstTokenx(),
+      $decl_without_final,
+    );
+
+    return $method->withFunctionDeclHeader($decl_without_final);
+  }
+
+  private static function preserveTrivia(
+    Token $final,
+    Token $first_token,
+    FunctionDeclarationHeader $context,
+  ): FunctionDeclarationHeader {
+
+    // This ->getCode() !== '' hack is here to give a nicer result in the common case.
+    // If your class looks like
+    // class Foo {
+    //   final public function bar(): void {}
+    // }      ^
+    // I would discard this space too, leaving public exactly where final used to be.
+    // If you actually have some useful trivia there (like a /*comment*/) I preserve it.
+
+    $final_trivia = NodeList::concat(
+      $final->getLeading(),
+      $final->getTrailing()
+        |> $$->getCode() !== ' ' ? $$ : new NodeList(),
+    );
+    $first_trivia = NodeList::concat($first_token->getLeading(), $final_trivia);
+
+    return $context->replaceDescendant(
+      $first_token,
+      $first_token->withLeading($first_trivia),
+    );
+  }
```

Comment:
does this rely on the assumption that `final` is the first modifier? what does it do for something like `public /* foo */ final /* bar */ function...`?

Reply:
```HACK
final class FinalPublicVsPublicFinal {
  /*1*/final/*2*/ public/*3*/ function fpf(): void {}
  /*1*/public/*2*/ final/*3*/ function pff(): void {}
}
final class FinalPublicVsPublicFinal {
  /*1*//*2*/ public/*3*/ function fpf(): void {}
  /*1*//*3*/ public/*2*/ function pff(): void {}
}
```

The trail of final by prepending it to the trivia of the first token (or the second if the first token is `final`). Since I don't check /where/ final was in relation to the other token, this might end up changing the order of the trivia. If this is a problem, I will have to write logic that deals with appending it after the trailing travia of the token, to keep the relative order.

Reply:
Hmm maybe I'm missing something but it seems to me that your solution is unnecessarily complicated. Why not just take all the trivia and prepend it to the next token (or append to previous -- doesn't matter)? Something like:

```
$next = $ast->getNextToken($final_token);
return $node->replace($next, $next->withLeading(Vec\concat($final_token->getLeading(), $final_token->getTrailing(), $next->getLeading())));
```

plus some code to preserve the "single space" edge case, but that shouldn't make it much more complicated (probably just replace `$final_token->getTrailing()` with the same ternary expression you used)